drivers/stepper/adi_tmc/adi_tmc5160_stepper_controller.c

```diff
@@ -0,0 +1,670 @@
+}
+
+static int tmc5160_stepper_set_target_position(const struct device *dev, const int32_t position)
+{
+	LOG_DBG("Stepper motor controller %s set target position to %d", dev->name, position);
+	const struct tmc5160_stepper_config *config = dev->config;
+	struct tmc5160_stepper_data *data = dev->data;
+	int err;
+
+	if (config->is_sg_enabled) {
+		stallguard_enable(dev, false);
+	}
+
+	err = tmc5160_write(config->controller, TMC5160_RAMPMODE,
+			    TMC5160_RAMPMODE_POSITIONING_MODE);
+	if (err != 0) {
+		return -EIO;
+	}
+	tmc5160_write(config->controller, TMC5160_XTARGET, position);
+
+	if (config->is_sg_enabled) {
+		k_work_reschedule(&data->stallguard_dwork,
+				  K_MSEC(config->sg_velocity_check_interval_ms));
+	}
+#ifdef CONFIG_STEPPER_ADI_TMC5160_RAMPSTAT_POLL
+	if (data->callback) {
+		k_work_reschedule(
+			&data->rampstat_callback_dwork,
+			K_MSEC(CONFIG_STEPPER_ADI_TMC5160_RAMPSTAT_POLL_INTERVAL_IN_MSEC));
+	}
+#endif
+	return 0;
+}
+
+static int tmc5160_stepper_enable_constant_velocity_mode(const struct device *dev,
+							 const enum stepper_direction direction,
+							 const uint32_t velocity)
+{
+	LOG_DBG("Stepper motor controller %s enable constant velocity mode", dev->name);
+	const struct tmc5160_stepper_config *config = dev->config;
+	struct tmc5160_stepper_data *data = dev->data;
+	uint32_t velocity_fclk;
+	int err;
+
+	calculate_velocity_from_hz_to_fclk(config->controller, velocity, &velocity_fclk);
+
+	if (config->is_sg_enabled) {
+		err = stallguard_enable(dev, false);
+		if (err != 0) {
+			return -EIO;
+		}
+	}
+
+	switch (direction) {
+	case STEPPER_DIRECTION_POSITIVE:
+		err = tmc5160_write(config->controller, TMC5160_RAMPMODE,
+				    TMC5160_RAMPMODE_POSITIVE_VELOCITY_MODE);
+		if (err != 0) {
+			return -EIO;
+		}
+		err = tmc5160_write(config->controller, TMC5160_VMAX, velocity_fclk);
+		if (err != 0) {
+			return -EIO;
+		}
+		break;
+
+	case STEPPER_DIRECTION_NEGATIVE:
+		err = tmc5160_write(config->controller, TMC5160_RAMPMODE,
+				    TMC5160_RAMPMODE_NEGATIVE_VELOCITY_MODE);
+		if (err != 0) {
+			return -EIO;
+		}
+		err = tmc5160_write(config->controller, TMC5160_VMAX, velocity_fclk);
+		if (err != 0) {
+			return -EIO;
+		}
+		break;
+	}
+
+	if (config->is_sg_enabled) {
+		k_work_reschedule(&data->stallguard_dwork,
+				  K_MSEC(config->sg_velocity_check_interval_ms));
+	}
+#ifdef CONFIG_STEPPER_ADI_TMC5160_RAMPSTAT_POLL
+	if (data->callback) {
+		k_work_reschedule(
+			&data->rampstat_callback_dwork,
+			K_MSEC(CONFIG_STEPPER_ADI_TMC5160_RAMPSTAT_POLL_INTERVAL_IN_MSEC));
+	}
+#endif
+	return 0;
+}
+
+#ifdef CONFIG_STEPPER_ADI_TMC_RAMP_GEN
+
+int tmc5160_stepper_set_ramp(const struct device *dev,
+			     const struct tmc_ramp_generator_data *ramp_data)
+{
+	LOG_DBG("Stepper motor controller %s set ramp", dev->name);
+	const struct tmc5160_stepper_config *config = dev->config;
+	int err;
+
+	err = tmc5160_write(config->controller, TMC5160_VSTART, ramp_data->vstart);
+	if (err != 0) {
+		return -EIO;
```

Comment:
do you have a reason to discard `err` and return `-EIO` instead? It's usually a good idea to pass through the error code, than one can log it and it may give some more information of what's wrong in the lower layers